### PR TITLE
HADOOP-16658 - S3A connector does not support including the token ren…

### PR DESCRIPTION
**Pull request by pzampino**
…ewer in the token identifier

Ran 'mvn verify' with endpoint s3.us-east-1.amazonaws.com
Two tests failed, but passed when run individually:
* ITestS3AConfiguration#testAutomaticProxyPortSelection
* ITestS3AInconsistency#testGetFileStatus

## NOTICE

Please create an issue in ASF JIRA before opening a pull request,
and you need to set the title of the pull request which starts with
the corresponding JIRA issue number. (e.g. HADOOP-XXXXX. Fix a typo in YYY.)
For more details, please see https://cwiki.apache.org/confluence/display/HADOOP/How+To+Contribute

### CODE REVIEW

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java [hunk 1]
```diff
@@ -3186,7 +3187,8 @@ public String getCanonicalServiceName() {
     entryPoint(Statistic.INVOCATION_GET_DELEGATION_TOKEN);
     LOG.debug("Delegation token requested");
     if (delegationTokens.isPresent()) {
-      return delegationTokens.get().getBoundOrNewDT(encryptionSecrets);
+      return delegationTokens.get().getBoundOrNewDT(encryptionSecrets,
+          (renewer!=null ? new Text(renewer) : null));
```
Comment (lmccay) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java]: Is it possible for renewer to not be null but actually be empty?

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/auth/delegation/AbstractDelegationTokenBinding.java [hunk 2]
```diff
@@ -157,17 +159,19 @@ public Text getOwnerText() {
...
    * @param encryptionSecrets encryption secrets for the token.
+   * @param renewer the principal permitted to renew the token.
    * @return the token data to include in the token identifier.
    * @throws IOException failure creating the token data.
    */
   public abstract AbstractS3ATokenIdentifier createTokenIdentifier(
       Optional<RoleModel.Policy> policy,
-      EncryptionSecrets encryptionSecrets) throws IOException;
+      EncryptionSecrets encryptionSecrets,
+      Text renewer) throws IOException;
```
Comment (lmccay) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/auth/delegation/AbstractDelegationTokenBinding.java]: As these are public methods, are you sure that we don't need to preserve the old signature as well for possible extensions? Does this appear in a public release yet at all?

Comment (pzampino) [reply hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/auth/delegation/AbstractDelegationTokenBinding.java]: This has not yet appeared in a public release; Steve actually suggested that I remove the previously-existing signatures, which I had marked as deprecated initially.

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java (hunk 1 above)
Comment (pzampino) [reply hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java]: It is possible for it to be empty, in which case the renewer Text object will be "empty", and that case is checked by yarn.

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/auth/delegation/AbstractDelegationTokenBinding.java (hunk 2 above)
Comment (steveloughran) [reply hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/auth/delegation/AbstractDelegationTokenBinding.java]: no ASF release. The only people using this outside the hadoop-aws jar are you; I have plans to do other incompatible changes soon. Sorry.

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/auth/delegation/TestS3ADelegationTokenSupport.java [hunk 3]
```diff
@@ -82,6 +85,7 @@ public void testSessionTokenDecode() throws Throwable {
     assertEquals("name of " + decodedUser,
         "alice",
         decodedUser.getUserName());
+    assertEquals(renewer, decoded.getRenewer());
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/auth/delegation/TestS3ADelegationTokenSupport.java]: add a message for better diags, e.g. "renewer"

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/auth/delegation/TestS3ADelegationTokenSupport.java [hunk 4]
```diff
@@ -97,9 +101,11 @@ public void testFullTokenKind() throws Throwable {
 
   @Test
   public void testSessionTokenIdentifierRoundTrip() throws Throwable {
+    Text renewer = new Text("yarn");
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/auth/delegation/TestS3ADelegationTokenSupport.java]: yarn@EXAMPLE.COM?

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/auth/delegation/TestS3ADelegationTokenSupport.java [hunk 5]
```diff
@@ -110,13 +116,34 @@ public void testSessionTokenIdentifierRoundTrip() throws Throwable {
     assertEquals("credentials in " + ids,
         id.getMarshalledCredentials(),
         result.getMarshalledCredentials());
+    assertEquals(renewer, id.getRenewer());
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/auth/delegation/TestS3ADelegationTokenSupport.java]: again, add a Message

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/auth/delegation/TestS3ADelegationTokenSupport.java [hunk 6]
```diff
@@ -110,13 +116,34 @@ public void testSessionTokenIdentifierRoundTrip() throws Throwable {
...
+        new MarshalledCredentials("a", "b", "c"),
+        new EncryptionSecrets(), "");
+
+    SessionTokenIdentifier result = S3ATestUtils.roundTrip(id, null);
+    String ids = id.toString();
+    assertEquals("URI in " + ids, id.getUri(), result.getUri());
+    assertEquals("credentials in " + ids,
+        id.getMarshalledCredentials(),
+        result.getMarshalledCredentials());
+    assertEquals(new Text(), id.getRenewer());
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/auth/delegation/TestS3ADelegationTokenSupport.java]: add message

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/auth/delegation/TestS3ADelegationTokenSupport.java [hunk 7]
```diff
@@ -142,6 +172,7 @@ public void testFullTokenIdentifierRoundTrip() throws Throwable {
     assertEquals("credentials in " + ids,
         id.getMarshalledCredentials(),
         result.getMarshalledCredentials());
+    assertEquals(renewer, result.getRenewer());
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/auth/delegation/TestS3ADelegationTokenSupport.java]: message `"renewer in " + ids`

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java (hunk 1 above)
Comment (steveloughran) [reply hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java]: * need to make sure that null results in renewer == new Text()
* add some spaces round the `!=`

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/auth/delegation/AbstractS3ATokenIdentifier.java [hunk 8]
```diff
@@ -103,17 +103,19 @@
...
    */
   protected AbstractS3ATokenIdentifier(
       final Text kind,
       final URI uri,
       final Text owner,
+      final Text renewer,
       final String origin,
       final EncryptionSecrets encryptionSecrets) {
-    this(kind, owner, new Text(), new Text(), uri);
+    this(kind, owner, renewer, new Text(), uri);
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/auth/delegation/AbstractS3ATokenIdentifier.java]: the renewer may now be null. should we add a `new Text()` if a null comes in?

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/auth/delegation/ITestSessionDelegationTokens.java [hunk 9]
```diff
@@ -118,7 +118,7 @@ public void testSaveLoadTokens() throws Throwable {
     EncryptionSecrets encryptionSecrets = new EncryptionSecrets(
         S3AEncryptionMethods.SSE_KMS, KMS_KEY);
     Token<AbstractS3ATokenIdentifier> dt
-        = delegationTokens.createDelegationToken(encryptionSecrets);
+        = delegationTokens.createDelegationToken(encryptionSecrets, null);
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/auth/delegation/ITestSessionDelegationTokens.java]: add an assert in this test case to verify load of renewer

**Comment by steveloughran** [comment]
Code LGTM; some minor comments about tests

I'd like AbstractS3ATokenIdentifier to create a Text() if a null renewer was passed in; this is consistent with the existing code.

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/auth/delegation/TestS3ADelegationTokenSupport.java (hunk 4 above)
Comment (pzampino) [reply hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/auth/delegation/TestS3ADelegationTokenSupport.java]: The value is the "short name"

**Comment by pzampino** [comment]
I've made the suggested changes, and re-ran the tests (which passed again).

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 1777 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 4 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1119 | trunk passed |
| +1 | compile | 38 | trunk passed |
| +1 | checkstyle | 28 | trunk passed |
| +1 | mvnsite | 40 | trunk passed |
| +1 | shadedclient | 820 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 30 | trunk passed |
| 0 | spotbugs | 60 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 59 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 33 | the patch passed |
| +1 | compile | 29 | the patch passed |
| +1 | javac | 29 | the patch passed |
| -0 | checkstyle | 20 | hadoop-tools/hadoop-aws: The patch generated 3 new + 9 unchanged - 0 fixed = 12 total (was 9) |
| +1 | mvnsite | 33 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 792 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 26 | the patch passed |
| +1 | findbugs | 70 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 86 | hadoop-aws in the patch passed. |
| +1 | asflicense | 33 | The patch does not generate ASF License warnings. |
| | | 5119 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.4 Server=19.03.4 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1664/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1664 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux f23e524c59ac 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / a901405 |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1664/2/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1664/2/testReport/ |
| Max. process+thread count | 472 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1664/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by steveloughran** [comment]
LGTM; thanks for changes

+1 and merged to trunk

**Pull request closed without merging**